src/app.js

```diff
@@ -53,32 +52,72 @@ export default function (options) {
 
 			if (typeof action === "function") {
 				container[key] = function (data) {
-					hooks.onAction(name, data)
+					hook("onAction", function (fn) {
+						fn(name, data)
+					})
+
+					if (shouldRender) {
+						var oldModel = model
+						model = merge(model, action(model, data))
+
+						hook("onUpdate", function (fn) {
+							fn(oldModel, model, data)
+						})
+
+						render(model, view)
 
-					if (shouldUpdate) {
-						hooks.onUpdate(model, model = merge(model, action(model, data)), data)
-						render(model, view, node)
 						return actions
 					} else {
-						return action(model, actions, data, hooks.onError)
+						return action(model, actions, data, onError)
 					}
 				}
 			} else {
-				initializeActions(container[key], action, shouldUpdate, name)
+				init(container[key], action, shouldRender, name)
 			}
 		})
 	}
 
-	function domContentLoaded(initApp) {
+	function load(fn) {
 		if (document.readyState[0] !== "l") {
-			initApp()
+			fn()
 		} else {
-			document.addEventListener("DOMContentLoaded", initApp)
+			document.addEventListener("DOMContentLoaded", fn)
 		}
 	}
 
-	function render(model, view, lastNode) {
-		patch(root, node = view(model, actions), lastNode, 0)
+	function use(app) {
+		if (app.model !== undefined) {
+			model = merge(model, app.model)
+		}
+
+		reducers = merge(reducers, app.reducers)
+		effects = merge(effects, app.effects)
+
+		if (app.subscriptions) {
+			subscriptions = subscriptions.concat(app.subscriptions)
+		}
+
+		if (app.hooks) {
+			hooks.push(app.hooks)
```

Comment:
Instead of pushing objects into an array I would iterate through the objects and construct this data-structure:

```javascript
var hooks = {
	onRender: [f1, f2, f3],
	onError: [f4]
}
```

This has the advantage that access is fast and easy. Basically it should be handled the same as `subscriptions`.